# Выбор локации для сквижины

<b>Цель:</b> определения региона, где добыча принесёт наибольшую прибыль.

<b>План работы:</b>

1) Загрузка и подготовка данных

2) Обучение и проверка модели для каждого региона

3) Подсчет рисков и прибыли для каждого региона

4) Вывод

# Загрузка и подготовка данных

In [1]:
BUDGET = 10_000_000_000
WELLS_COUNT = 200
REVENUE_PER_UNIT = 450_000

In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [3]:
geo_data_0 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_0.csv')
geo_data_1 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_1.csv')
geo_data_2 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_2.csv')


In [4]:
geo_data_0.head()

,id,f0,f1,f2,product
0,txEyH,0.705745,-0.497823,1.221170,105.280062
1,2acmU,1.334711,-0.340164,4.365080,73.037750
2,409Wp,1.022732,0.151990,1.419926,85.265647
3,iJLyR,-0.032172,0.139033,2.978566,168.620776
4,Xdl7t,1.988431,0.155413,4.751769,154.036647


In [5]:
geo_data_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [6]:
geo_data_0.duplicated().sum()

np.int64(0)

In [7]:
geo_data_1.head()

,id,f0,f1,f2,product
0,kBEdx,-15.001348,-8.276000,-0.005876,3.179103
1,62mP7,14.272088,-3.475083,0.999183,26.953261
2,vyE1P,6.263187,-5.948386,5.001160,134.766305
3,KcrkZ,-13.081196,-11.506057,4.999415,137.945408
4,AHL4O,12.702195,-8.147433,5.004363,134.766305


In [8]:
geo_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [9]:
geo_data_1.duplicated().sum()

np.int64(0)

In [10]:
geo_data_2.head()

,id,f0,f1,f2,product
0,fwXo0,-1.146987,0.963328,-0.828965,27.758673
1,WJtFt,0.262778,0.269839,-2.530187,56.069697
2,ovLUW,0.194587,0.289035,-5.586433,62.871910
3,q6cA6,2.236060,-0.553760,0.930038,114.572842
4,WPMUX,-0.515993,1.716266,5.899011,149.600746


In [11]:
geo_data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [12]:
geo_data_2.duplicated().sum()

np.int64(0)

<div class="alert alert-info">

**Вывод:** 
- Датасет содержит 100000 записей
- 5 столбцов: id, f0, f1, f2 (признаки), product (целевая переменная)
- Пропусков: 0
- Дубликатов: 0
- Все данные числовые, готовы для обучения модели
</div>

# Обучение и проверка модели для каждого региона

In [13]:
X = geo_data_0.drop(['product', 'id'], axis=1)
y = geo_data_0['product']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=12345)

model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

pred_0 = pd.Series(y_pred)
test_0 = y_test.reset_index(drop=True)

print('mean_product', y_pred.mean())
print('RMSE:', mean_squared_error(y_test, y_pred) ** 0.5)

mean_product 92.59256778438038
RMSE: 37.5794217150813


In [14]:
X = geo_data_1.drop(['product', 'id'], axis=1)
y = geo_data_1['product']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=12345)

model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

pred_1 = pd.Series(y_pred)
test_1 = y_test.reset_index(drop=True)

print('mean_product', y_pred.mean())
print('RMSE:', mean_squared_error(y_test, y_pred) ** 0.5)

mean_product 68.728546895446
RMSE: 0.8930992867756165


In [15]:
X = geo_data_2.drop(['product', 'id'], axis=1)
y = geo_data_2['product']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=12345)

model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

pred_2 = pd.Series(y_pred)
test_2 = y_test.reset_index(drop=True)

print('mean_product', y_pred.mean())
print('RMSE:', mean_squared_error(y_test, y_pred) ** 0.5)

mean_product 94.96504596800489
RMSE: 40.02970873393434


<div class="alert alert-info">

**Анализ результатов по трём регионам:**

- **Регион 0:** Средний запас 92.6 тыс. баррелей, RMSE = 37.6. Высокая вариативность запасов.
- **Регион 1:** Средний запас 68.7 тыс. баррелей, RMSE = 0.89. Модель предсказывает очень точно — низкий разброс.
- **Регион 2:** Средний запас 95.0 тыс. баррелей, RMSE = 40.0. Самый высокий средний запас, но максимальная ошибка предсказания.

**Предварительный вывод:** 
Регион 2 имеет наибольший средний запас, но высокая ошибка модели создаёт риски. 
Регион 1 показывает самые точные предсказания, но низкий средний запас. 
Окончательное решение примем после расчёта прибыли и рисков (этап 5).
</div>

# Подсчет рисков и прибыли для каждого региона

In [16]:
budget_per_well = BUDGET / WELLS_COUNT
print('Бюджет на одну сквижну:', budget_per_well)

breakeven_volume = budget_per_well / REVENUE_PER_UNIT
print('Безубыточный объем для одной скважины:', breakeven_volume, 'тыс. баррелей')

print('Регион 0:', pred_0.mean(), 'тыс. бараллей')
print('Регион 1:', pred_1.mean(), 'тыс. бараллей')
print('Регион 2:', pred_2.mean(), 'тыс. бараллей')

Бюджет на одну сквижну: 50000000.0
Безубыточный объем для одной скважины: 111.11111111111111 тыс. баррелей
Регион 0: 92.59256778438038 тыс. бараллей
Регион 1: 68.728546895446 тыс. бараллей
Регион 2: 94.96504596800489 тыс. бараллей


<div class="alert alert-info">

**Выводы:**

- Безубыточный объем для одной скважины 111.1 тыс. бараллей.
- Средний запас в регионах ниже безубыточного.
- Нужно выбрать 200 лучших скважин чтобы получить прибыль.

</div>

In [17]:
def revenue(ans, pred, count=200):
    top_count = pred.argsort()[-count:]
    total_volume = ans.iloc[top_count].sum()
    return total_volume * REVENUE_PER_UNIT - BUDGET

In [19]:
print('Максимальная прибыль для Региона 0:', revenue(test_0, pred_0))
print('Максимальная прибыль для Региона 1:', revenue(test_1, pred_1))
print('Максимальная прибыль для Региона 2:', revenue(test_2, pred_2))

Максимальная прибыль для Региона 0: 3320826043.1398506
Максимальная прибыль для Региона 1: 2415086696.681511
Максимальная прибыль для Региона 2: 2710349963.5998325


<div class="alert alert-info">

**Выводы:**

- **Регион 0:** является лидером по прибыли, которая составляет 3.3 млрд. руб.

- **Регион 1:** является хужшим по прибыли, которая составляет 2.4 млрд. руб.

- **Регион 1:** является средним по прибыли, которая составляет 2.7 млрд. руб.

</div>

In [24]:
def bootstrap(y_test, y_pred, n=1000):
    state = np.random.RandomState(12345)
    profits = []

    for i in range(n):
        subsample = y_pred.sample(n=500, replace=True, random_state=state)
        pred_sorted = subsample.sort_values(ascending=False)
        selected = y_test[pred_sorted.index][:200]
        total_volume = selected.sum()
        profit = total_volume * REVENUE_PER_UNIT - BUDGET
        profits.append(profit)

    return pd.Series(profits)

<div style="border:solid blue 2px; padding: 20px">

<div class="alert alert-info">

Мы берем выборки с повторениями? Почему?

In [34]:
profits_0 = bootstrap(test_0, pred_0)

mean_profit = profits_0.mean()
ci_lower = profits_0.quantile(0.025)
ci_upper = profits_0.quantile(0.975)
risk = (profits_0 < 0).mean() * 100

print(f'Средняя прибыль: {mean_profit:,.0f} руб.')
print(f'95% ДИ: [{ci_lower:,.0f}, {ci_upper:,.0f}] руб.')
print(f'Риск убытков: {risk:.2f}%')

Средняя прибыль: 396,164,985 руб.
95% ДИ: [-111,215,546, 909,766,942] руб.
Риск убытков: 6.90%


In [33]:
profits_1 = bootstrap(test_1, pred_1)

mean_profit = profits_1.mean()
ci_lower = profits_1.quantile(0.025)
ci_upper = profits_1.quantile(0.975)
risk = (profits_1 < 0).mean() * 100

print(f'Средняя прибыль: {mean_profit:,.0f} руб.')
print(f'95% ДИ: [{ci_lower:,.0f}, {ci_upper:,.0f}] руб.')
print(f'Риск убытков: {risk:.2f}%')

Средняя прибыль: 456,045,106 руб.
95% ДИ: [33,820,509, 852,289,454] руб.
Риск убытков: 1.50%


In [31]:
profits_2 = bootstrap(test_2, pred_2)

mean_profit = profits_2.mean()
ci_lower = profits_2.quantile(0.025)
ci_upper = profits_2.quantile(0.975)
risk = (profits_2 < 0).mean() * 100

print(f'Средняя прибыль: {mean_profit:,.0f} руб.')
print(f'95% ДИ: [{ci_lower:,.0f}, {ci_upper:,.0f}] руб.')
print(f'Риск убытков: {risk:.2f}%')

Средняя прибыль: 404,403,867 руб.
95% ДИ: [-163,350,413, 950,359,575] руб.
Риск убытков: 7.60%


<div class="alert alert-info">

**Итоговые выводы:**

После применения техники Bootstrap с 1000 выборок и анализа рисков получены следующие результаты:

**Регион 0:** 
- Средняя прибыль: 396 млн руб.
- 95% доверительный интервал: [-111 млн, 910 млн] руб.
- Риск убытков: 6.90%
- **Вывод:** Не подходит — риск убытков превышает допустимый порог 2.5%

**Регион 1:** 
- Средняя прибыль: 456 млн руб.
- 95% доверительный интервал: [34 млн, 852 млн] руб.
- Риск убытков: 1.50%
- **Вывод:** ПОДХОДИТ — риск убытков менее 2.5%

**Регион 2:** 
- Средняя прибыль: 404 млн руб.
- 95% доверительный интервал: [-163 млн, 950 млн] руб.
- Риск убытков: 7.60%
- **Вывод:** Не подходит — риск убытков превышает допустимый порог 2.5%

---

**Рекомендация:** Выбрать **Регион 1** для разработки скважин.

**Обоснование:**
1. Риск убытков составляет 1.5%, что соответствует требованию компании (менее 2.5%)
2. Средняя прибыль 456 млн руб. — максимальная среди подходящих регионов
3. Модель показывает высокую точность предсказаний (RMSE = 0.89)
4. Узкий доверительный интервал указывает на стабильность прибыли
5. Нижняя граница 95% ДИ положительная (34 млн руб.) — минимальный риск убытков

Регионы 0 и 2 не рекомендуются для разработки из-за высокого риска убытков (6.9% и 7.6% соответственно), несмотря на потенциально высокую прибыль.

</div>